core/src/test/java/org/apache/accumulo/core/file/rfile/MultiThreadedRFileTest.java

```diff
@@ -0,0 +1,387 @@
+        }
+        StringWriter string = new StringWriter();
+        PrintWriter writer = new PrintWriter(string);
+        t.printStackTrace(writer);
+        writer.flush();
+        stackTrace.put(msg, string.getBuffer().toString());
+      }
+    } finally {
+      trfBase.closeReader();
+      trfBase.close();
+    }
+
+    for (String message : messages.keySet()) {
+      System.out.println(messages.get(message) + ": " + message);
+      System.out.println(stackTrace.get(message));
+    }
+
+    assertTrue(threadExceptions.isEmpty());
+  }
+
+  private void validate(TestRFile trf) throws IOException {
+    Random random = new Random();
+    for (int iteration = 0; iteration < 10; iteration++) {
+      int part = random.nextInt(4);
+
+      Range range = new Range(getKey(part, 0, 0), true, getKey(part, 4, 2048), true);
+      trf.iter.seek(range, EMPTY_COL_FAMS, false);
+
+      Key last = null;
+      for (int locality = 0; locality < 4; locality++) {
+        for (int i = 0; i < 2048; i++) {
+          Key key = getKey(part, locality, i);
+          Value value = getValue(i);
+          assertTrue("No record found for row " + part + " locality " + locality + " index " + i, trf.iter.hasTop());
+          assertEquals("Invalid key found for row " + part + " locality " + locality + " index " + i, key, trf.iter.getTopKey());
+          assertEquals("Invalie value found for row " + part + " locality " + locality + " index " + i, value, trf.iter.getTopValue());
+          last = trf.iter.getTopKey();
+          trf.iter.next();
+        }
+      }
+      if (trf.iter.hasTop()) {
+        assertFalse("Found " + trf.iter.getTopKey() + " after " + last + " in " + range, trf.iter.hasTop());
+      }
+
+      range = new Range(getKey(4, 4, 0), true, null, true);
+      trf.iter.seek(range, EMPTY_COL_FAMS, false);
+      if (trf.iter.hasTop()) {
+        assertFalse("Found " + trf.iter.getTopKey() + " in " + range, trf.iter.hasTop());
+      }
+    }
+
+    Range range = new Range((Key) null, null);
+    trf.iter.seek(range, EMPTY_COL_FAMS, false);
+
+    Key last = null;
+    for (int part = 0; part < 4; part++) {
+      for (int locality = 0; locality < 4; locality++) {
+        for (int i = 0; i < 2048; i++) {
+          Key key = getKey(part, locality, i);
+          Value value = getValue(i);
+          assertTrue("No record found for row " + part + " locality " + locality + " index " + i, trf.iter.hasTop());
+          assertEquals("Invalid key found for row " + part + " locality " + locality + " index " + i, key, trf.iter.getTopKey());
+          assertEquals("Invalie value found for row " + part + " locality " + locality + " index " + i, value, trf.iter.getTopValue());
+          last = trf.iter.getTopKey();
+          trf.iter.next();
+        }
+      }
+    }
+
+    if (trf.iter.hasTop()) {
+      assertFalse("Found " + trf.iter.getTopKey() + " after " + last + " in " + range, trf.iter.hasTop());
+    }
```

Comment:
Since the test created locality groups, could have another loop that seeks w/ a non-empty set of column families.  I don't think this has to be done, its just something I thought of while looking at the test.